File: app/src/main/java/com/zofers/zofers/ui/notifications/messenger/MessageBaseViewHolder.java
package com.zofers.zofers.ui.notifications.messenger;

import android.text.method.LinkMovementMethod;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.zofers.zofers.R;
import com.zofers.zofers.model.Conversation;
import com.zofers.zofers.model.Message;
import com.zofers.zofers.model.ObjectId;
import com.zofers.zofers.model.Participant;
import com.zofers.zofers.staff.DateTimeUtils;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import androidx.recyclerview.widget.RecyclerView;

public abstract class MessageBaseViewHolder extends RecyclerView.ViewHolder {
    private static Map<String, Object> attachmentStore = new HashMap<>();
    public ViewGroup attachmentContainer;
    public TextView messageTextView;
    public TextView timeTextView;
    public TextView seenTextView;
    public RecyclerView seenHeadsRecyclerView;
    public View topSpace;
    public View bottomSpace;

    public MessageBaseViewHolder(View itemView) {
        super(itemView);
        messageTextView = itemView.findViewById(R.id.message_textView);
        timeTextView = itemView.findViewById(R.id.message_time_textView);
        seenTextView = itemView.findViewById(R.id.message_seen_textView);
        topSpace = itemView.findViewById(R.id.top_space);
        bottomSpace = itemView.findViewById(R.id.bottom_space);
        messageTextView.setMovementMethod(LinkMovementMethod.getInstance());
    }

    public void bind(Message message, int messageType, boolean showTime, Conversation conversation, String profileID) {
        messageTextView.setText(message.getText());
        makeMessageBaloonForm(messageType, showTime);
//        //seen logic
//        List<Participant> participants = conversation.getParticipantsExcept(profileID);
//        if (participants.size() > 0) {
//            String lastSeenId = participants.get(0).getLastSeenMessageId();
//            if (profileID == message.getUserId() && (lastSeenId == null || new ObjectId(message.getId()).compareTo(new ObjectId(lastSeenId)) == 1)) {
//                seenTextView.setText(seenTextView.getContext().getString(R.string.messenger_sent));
//            } else {
//                seenTextView.setText(seenTextView.getContext().getString(R.string.messenger_seen));
//            }
//        }
//
//        seenTextView.setVisibility(seenTextDefaultVisibility(message));
//        //end seen logic
        if (message.getDate() != null) {
            timeTextView.setText(DateTimeUtils.getSmartDate(message.getDate(), false));
            timeTextView.setVisibility(showTime ? View.VISIBLE : View.GONE);
        }
    }

    public abstract void makeMessageBaloonForm(int messageType, boolean showTime);

    public abstract int seenTextDefaultVisibility(Message message);
}

File: app/src/main/java/com/zofers/zofers/ui/notifications/messenger/MessageBaseViewHolder.kt
package com.zofers.zofers.ui.notifications.messenger

import android.text.method.LinkMovementMethod
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.zofers.zofers.R
import com.zofers.zofers.model.Conversation
import com.zofers.zofers.model.Message
import com.zofers.zofers.model.ObjectId
import com.zofers.zofers.model.Participant
import com.zofers.zofers.staff.DateTimeUtils
import java.util.*

abstract class MessageBaseViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
	var attachmentContainer: ViewGroup
	var messageTextView: TextView
	var timeTextView: TextView
	var seenTextView: TextView
	var topSpace: View
	var bottomSpace: View


	fun bind(message: Message, messageType: Int, showTime: Boolean, conversation: Conversation?, profileID: String?) {
		messageTextView.text = message.text
		makeMessageBaloonForm(messageType, showTime)
		//        //seen logic
//		val participants = conversation?.getParticipantsExcept(profileID);
//		if (!participants.isNullOrEmpty()) {
//			val lastSeenId = participants[0].lastSeenMessageId;
//			if (profileID == message.userId && ObjectId(message.id).compareTo(ObjectId(lastSeenId)) == 1) {
//				seenTextView.text = seenTextView.context.getString(R.string.messenger_sent);
//			} else {
//				seenTextView.text = seenTextView.context.getString(R.string.messenger_seen);
//			}
//		}
//
//		seenTextView.visibility = seenTextDefaultVisibility(message);
//        //end seen logic
		if (message.date != null) {
			timeTextView.text = DateTimeUtils.getSmartDate(message.date, false)
			timeTextView.visibility = if (showTime) View.VISIBLE else View.GONE
		}
	}

	abstract fun makeMessageBaloonForm(messageType: Int, showTime: Boolean)
	abstract fun seenTextDefaultVisibility(message: Message): Int

	companion object {
		private val attachmentStore: Map<String, Any> = HashMap()
	}

	init {
		messageTextView = itemView.findViewById(R.id.message_textView)
		timeTextView = itemView.findViewById(R.id.message_time_textView)
		seenTextView = itemView.findViewById(R.id.message_seen_textView)
		topSpace = itemView.findViewById(R.id.top_space)
		bottomSpace = itemView.findViewById(R.id.bottom_space)
		attachmentContainer = itemView.findViewById(R.id.attachment_container)
		messageTextView.movementMethod = LinkMovementMethod.getInstance()
	}
}